startup benchmarks with Native Image

Akirathan opened the pull request:
Closes #13291

### Pull Request Description

Introduce `Native_Image_Startup` benchmarks. The old `Startup` benchmarks are kept intact to preserve history. 

### Important Notes

[Benchmark Standard Libraries GH workflow](https://github.com/enso-org/enso/actions/runs/15873473973/workflow) now builds NI with `ENSO_LAUNCHER=native`. See changes to [engine.rs](https://github.com/enso-org/enso/pull/13360/files#diff-ddb74013689698a7a9da05e81c3edcfb0ddd7fa8015f03546ea5476d3260acdf).

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.
- [x] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

Commit:
add Native_Image_Startup benchmarks

Commit:
runs_in_aot check is deferred to the start of the bench execution.

Otherwise, bench compilation `std-benchmarks/compile` fails

Commit:
EnsoJMH benchmarks require native engine runner

Akirathan commented:
JVM and NI Startup benchs scheduled at https://github.com/enso-org/enso/actions/runs/15856989803

<blockquote><img src="https://opengraph.githubassets.com/23bb906c586b58b14798e43bf1ab2c6be6a61995c42468ea2ecab447b5d881b8/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Benchmark Standard Libraries · enso-org/enso@c16bf0a</a></strong></div><div>Enso Analytics is a self-service data prep and analysis platform designed for data teams. - Benchmark Standard Libraries · enso-org/enso@c16bf0a</div></blockquote>

**EDIT:**
It failed on the check whether running in AOT. This check was removed in https://github.com/enso-org/enso/pull/13360/commits/b0d6c08300cb86156d635baa90092f578029c796

JaroslavTulach commented on test/Benchmarks/src/Startup/Startup.enso:
```diff
@@ -35,20 +45,31 @@ type Data
-        group_builder.specify "hello_world_startup" data.bench_hello
-        group_builder.specify "import_world_startup" data.bench_import
+    [["Native_Image_Startup", False], ["JVM_Startup", True]].each \entry->
+        group_name = entry.at 0
```
```suggestion
        group_name = entry.first
```

Akirathan replied:
Applied in 5387fcdc57f4691497f0a5d7c777e8cc9a452cdf

JaroslavTulach commented on test/Benchmarks/src/Startup/Startup.enso:
```diff
@@ -35,20 +45,31 @@ type Data
-        group_builder.specify "import_world_startup" data.bench_import
+    [["Native_Image_Startup", False], ["JVM_Startup", True]].each \entry->
+        group_name = entry.at 0
+        in_jvm = (entry.at 1):Boolean
```
```suggestion
        in_jvm = (entry.second):Boolean
```

Akirathan replied:
Applied in 5387fcdc57f4691497f0a5d7c777e8cc9a452cdf

JaroslavTulach commented on test/Benchmarks/src/Startup/Startup.enso:
```diff
@@ -8,16 +8,26 @@ polyglot java import java.io.File as Java_File
+    bench_import self in_jvm:Boolean =
+        cmd = ["--run", self.import_world.to_text]
+        case in_jvm of
+            True -> self.startup (["--jvm"] + cmd)
```
- adding `in_jvm` argument to `self.startup` might remove some duplicated code

Akirathan replied:
Done in 4ae5572be2b368be986c9f8ffbbdf551edd7c8e1

JaroslavTulach commented on test/Benchmarks/src/Startup/Startup.enso:
```diff
@@ -35,20 +45,31 @@ type Data
-        group_builder.specify "empty_startup" data.bench_empty
-        group_builder.specify "hello_world_startup" data.bench_hello
-        group_builder.specify "import_world_startup" data.bench_import
+    [["Native_Image_Startup", False], ["JVM_Startup", True]].each \entry->
```
- It would be good to **not rename** the already executed benchmarks 
- otherwise the history is going to disconnect
- does that imply to keep "Startup" name for JVM launch?
- and introduce "Native Startup" for the new _native image_ launches?

Akirathan replied:
Fair point. `JVM_Startup` renamed back to just `Startup` in 531de8ce40aaa6d2560648cdb9981217c5df3f84 . So we will have:
- `Startup` which will run in JVM
- `Native_Image_Startup` which will run in NI.

JaroslavTulach reviewed (approved):
- try to keep continuity of already collected results
- (by keeping the name of JVM benchmarks)

Commit:
Do not rename already existing Startup benchmark

Commit:
Use first and second methods instead of indexes

Commit:
Move in_jvm parameter handling to the startup method

Commit:
Do not check if running in AOT.

This was only causing failures in `std-benchmarks/bench` on CI. Let's just sanity check the results.

Akirathan commented:
Another run of the startup benchmarks in https://github.com/enso-org/enso/actions/runs/15873473973

<blockquote><img src="https://opengraph.githubassets.com/5537df4117e97ac675c3da2e0721353b24ae2dac59bed870682289c27e1dfefd/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Benchmark Standard Libraries · enso-org/enso@b0d6c08</a></strong></div><div>Enso Analytics is a self-service data prep and analysis platform designed for data teams. - Benchmark Standard Libraries · enso-org/enso@b0d6c08</div></blockquote>

Akirathan commented:
`Native_Image_Startup.import_world_startup` is [failing](https://github.com/enso-org/enso/actions/runs/15873473973/job/44755413852#step:10:3507) on errors like
```
/runner/_work/enso/enso/test/built-distribution/enso-engine-2025.2.1-dev-linux-amd64/enso-2025.2.1-dev/lib/Standard/Database/2025.2.1-dev/src/SQL.enso:7:1: error: Package containing the module Standard.Database.SQL_Statement.SQL_Statement could not be loaded: The package could not be loaded: null
```

This looks like some configuration error in how the NI is built. Running `env ENSO_LAUNCHER=native sbt runEngineDistribution --run test/Benchmarks/src/Startup/Import_World.enso` works just fine.

**EDIT**
Turns out that on CI, `ENSO_LAUNCHER` was set to `native,test`. And that caused `Native_Image_Startup.import_world` benchmark to fail. I could reproduce that locally. Fixed in https://github.com/enso-org/enso/pull/13360/commits/beb3152a2566c2aee1a3df7c81c5b9d4061ab69c

Commit:
NI for benchmarks is built with ENSO_LAUNCHER=native,-ls

And not with ENSO_LAUNCHER=native,test. That caused failures in `Native_Image_Startup.import_world` benchmark.

Akirathan commented:
Third launch of `.*Startup.*` benchmarks scheduled in https://github.com/enso-org/enso/actions/runs/15883522804

<blockquote><img src="https://opengraph.githubassets.com/5537df4117e97ac675c3da2e0721353b24ae2dac59bed870682289c27e1dfefd/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Benchmark Standard Libraries · enso-org/enso@beb3152</a></strong></div><div>Enso Analytics is a self-service data prep and analysis platform designed for data teams. - Benchmark Standard Libraries · enso-org/enso@beb3152</div></blockquote>

Akirathan commented:
All the `.*Startup.*` benchmarks [succeeded](https://github.com/enso-org/enso/actions/runs/15883522804/job/44789701102#step:10:4093):
```
 INFO ide_ci::program::command: sbt ℹ️ [info] Benchmark                                  Mode  Cnt     Score      Error  Units
 INFO ide_ci::program::command: sbt ℹ️ [info] Native_Image_Startup.empty_startup         avgt    3    63.426 ±   12.988  ms/op
 INFO ide_ci::program::command: sbt ℹ️ [info] Native_Image_Startup.hello_world_startup   avgt    3   774.341 ±  122.012  ms/op
 INFO ide_ci::program::command: sbt ℹ️ [info] Native_Image_Startup.import_world_startup  avgt    3  2262.154 ±  380.344  ms/op
 INFO ide_ci::program::command: sbt ℹ️ [info] Startup.empty_startup                      avgt    3  2070.110 ±  375.887  ms/op
 INFO ide_ci::program::command: sbt ℹ️ [info] Startup.hello_world_startup                avgt    3  3105.051 ± 1328.718  ms/op
 INFO ide_ci::program::command: sbt ℹ️ [info] Startup.import_world_startup               avgt    3  3752.452 ± 1113.809  ms/op
```

At the end, we have to build the NI for benchmarks with `ENSO_LAUNCHER=native,-ls`, because `ENSO_LAUNCHER=native,test` does not work. See [my previous comment](https://github.com/enso-org/enso/pull/13360#issuecomment-3004841951).

JaroslavTulach commented:
- `Native_Image_Startup.hello_world_startup` is at **660ms**
- `Native_Image_Startup.import_world_startup` is at **1800ms**
- that is 600ms and 1700ms longer than it should be and shows how bad Enso compiler system is
